<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

## Lab: Cleaning Rock Song Data

_Authors: Dave Yerrington (SF)_

---


In [1]:
import numpy as np
import pandas as pd

### 1. Load `rock.csv` and do an initial examination of its data columns.

In [ ]:
# your answer here

In [2]:
pd.read_csv(r"/Users/PRSmb/OneDrive/General-Assembly/my-1019-repo/ClassMaterial/Unit2/data/rock.csv")

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Song Clean    2230 non-null   object
 1   ARTIST CLEAN  2230 non-null   object
 2   Release Year  1653 non-null   object
 3   COMBINED      2230 non-null   object
 4   First?        2230 non-null   int64 
 5   Year?         2230 non-null   int64 
 6   PlayCount     2230 non-null   int64 
 7   F*G           2230 non-null   int64 
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


In [6]:
df.head(5)

Song Clean ARTIST CLEAN Release Year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special          NaN   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                                COMBINED  First?  Year?  PlayCount  F*G  
0        Caught Up in You by .38 Special       1      1         82   82  
1            Fantasy Girl by .38 Special       1      0          3    0  
2         Hold On Loosely by .38 Special       1      1         85   85  
3  Rockin' Into the Night by .38 Special       1      1         18   18  
4              Art For Arts Sake by 10cc       1      1          1    1

In [16]:
?pd.Series.value_counts

In [17]:
df['Year?'].value_counts()

1    1653
0     577
Name: Year?, dtype: int64

In [18]:
(df['First?']).value_counts()

1    2230
Name: First?, dtype: int64

### 2.  Clean up the column names.

Let's clean up the column names. There are two ways we can accomplish this:

#### 2.A Change the column names when you import the data using `pd.read_csv()`.

Notice that, when passing `names=[..A LIST OF STRING..]` with a number of columns that matches the number of strings in the passed list, you replace the column names.

NOTE: When you create custom column names, the first row of the `.csv` already represents a header. It is important to tell `pandas` to skip that row. The `skiprows=1` keyword argument to `read_csv()` will tell `pandas` to skip the first row.

In [ ]:
# your answer here

In [20]:
clean_columns = [
    "song"
    ,"artist"
    ,"release_year"
    ,"song_artist"
    ,"first_flag"
    ,"year_flag"
    ,"playcount"
    ,'year_x_playcount'
]

In [22]:
print(clean_columns)

['song', 'artist', 'release_year', 'song_artist', 'first_flag', 'year_flag', 'playcount', 'year_x_playcount']


In [23]:
?pd.read_csv

In [111]:
df_a = pd.read_csv(
            filepath_or_buffer=r"/Users/PRSmb/OneDrive/General-Assembly/my-1019-repo/ClassMaterial/Unit2/data/rock.csv"
            ,names=clean_columns
            ,skiprows=1
           )

In [112]:
df_a.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   song              2230 non-null   object
 1   artist            2230 non-null   object
 2   release_year      1653 non-null   object
 3   song_artist       2230 non-null   object
 4   first_flag        2230 non-null   int64 
 5   year_flag         2230 non-null   int64 
 6   playcount         2230 non-null   int64 
 7   year_x_playcount  2230 non-null   int64 
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


In [113]:
df_a.head()

song       artist release_year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special          NaN   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                             song_artist  first_flag  year_flag  playcount  \
0        Caught Up in You by .38 Special           1          1         82   
1            Fantasy Girl by .38 Special           1          0          3   
2         Hold On Loosely by .38 Special           1          1         85   
3  Rockin' Into the Night by .38 Special           1          1         18   
4              Art For Arts Sake by 10cc           1          1          1   

   year_x_playcount  
0                82  
1                 0  
2                85  
3                18  
4                 1

#### 2.B Change column names using the `.rename()` function.

The `.rename()` function takes an argument, `columns=name_dict`, in which `name_dict` is a dictionary containing the original column names as keys and the new column names as values.

In [ ]:
# your answer here

In [37]:
clean_columns = [
    "song"
    ,"artist"
    ,"release_year"
    ,"song_artist"
    ,"first_flag"
    ,"year_flag"
    ,"playcount"
    ,'year_x_playcount'
]

In [38]:
orig_cols = list(df.columns)

In [48]:
clean_dict = {}
for i, j in zip(orig_cols, clean_columns):
    clean_dict.update({i:j})

In [49]:
clean_dict

{'Song Clean': 'song',
 'ARTIST CLEAN': 'artist',
 'Release Year': 'release_year',
 'COMBINED': 'song_artist',
 'First?': 'first_flag',
 'Year?': 'year_flag',
 'PlayCount': 'playcount',
 'F*G': 'year_x_playcount'}

In [50]:
df_b = df.rename(columns = clean_dict)

In [54]:
list(df.columns)

['Song Clean',
 'ARTIST CLEAN',
 'Release Year',
 'COMBINED',
 'First?',
 'Year?',
 'PlayCount',
 'F*G']

In [55]:
list(df_b.columns)

['song',
 'artist',
 'release_year',
 'song_artist',
 'first_flag',
 'year_flag',
 'playcount',
 'year_x_playcount']

In [110]:
df_b.head()

song       artist release_year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special            0   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                             song_artist  first_flag  year_flag  playcount  \
0        Caught Up in You by .38 Special           1          1         82   
1            Fantasy Girl by .38 Special           1          0          3   
2         Hold On Loosely by .38 Special           1          1         85   
3  Rockin' Into the Night by .38 Special           1          1         18   
4              Art For Arts Sake by 10cc           1          1          1   

   year_x_playcount  
0                82  
1                 0  
2                85  
3                18  
4                 1

#### 2.C Reassigning the `.columns` attribute of a DataFrame.

You can also just reassign the `.columns` attribute to a list of strings containing the new column names. 

The only caveat with reassigning `.columns` is that you have to reassign all of the column names at once. You can't partially replace a value by working on `.columns` directly. You have to reassign `.columns` with a list of equal length. 

In [ ]:
# your answer here

In [56]:
df_c = df

In [57]:
df_c.columns = clean_columns

In [59]:
df_c.head(2)

song       artist release_year  \
0  Caught Up in You  .38 Special         1982   
1      Fantasy Girl  .38 Special          NaN   

                       song_artist  first_flag  year_flag  playcount  \
0  Caught Up in You by .38 Special           1          1         82   
1      Fantasy Girl by .38 Special           1          0          3   

   year_x_playcount  
0                82  
1                 0

### 3. Subsetting data where null values exist.

We have mixed `str` and `NaN` values in the `release` column. `NaN` stands for "not a number" and is the way `pandas` handles "nulls" or nonexistent data. We can use the `.isnull()` method of a Series to find null values.

Print the header of the data subset to where the `release` column is null values.

In [ ]:
# your answer here

In [67]:
df_a[(df_a["release_year"].isnull())].head()

song       artist release_year  \
1             Fantasy Girl  .38 Special          NaN   
10   Baby, Please Don't Go        AC/DC          NaN   
13  CAN'T STOP ROCK'N'ROLL        AC/DC          NaN   
16        Girls Got Rhythm        AC/DC          NaN   
24         Let's Get It Up        AC/DC          NaN   

                        song_artist  first_flag  year_flag  playcount  \
1       Fantasy Girl by .38 Special           1          0          3   
10   Baby, Please Don't Go by AC/DC           1          0          1   
13  CAN'T STOP ROCK'N'ROLL by AC/DC           1          0          5   
16        Girls Got Rhythm by AC/DC           1          0         24   
24         Let's Get It Up by AC/DC           1          0          4   

    year_x_playcount  
1                  0  
10                 0  
13                 0  
16                 0  
24                 0

### 4. Update slices of your DataFrame based on mask selection/slices.

In many scenarios, we want to upate values in our DataFrame according to criteria. Let's say we wanted to set all of the null values in `release` to 0.

With newer versions of `pandas`, in order to manipulate data in the original DataFrame, we have to use `.loc` while performing reassignment using a mask and an index.

For example, the following won't always work:
```python
df[row_mask]['column_name'] = new_value
```

The best way to accomplish the same task is:
```python
df.loc[row_mask, 'column_name'] = new_value
```

For multiple column assignment, you would use:
```python
df.loc[row_mask, ['col_1', 'col_2', 'col_3']] = new_value
```

#### 4.A Let's try it out. Make all of the null values in `release` 0.

In [ ]:
# your answer here

# In the above "row_mask" means some sort of condition that evaluates to true or false.

In [108]:
df_a = df_b # Reset df_a

In [95]:
# This method works using the np where statement, but not the loc statemnt.
df_a['release_year'] = np.where(df_a["release_year"].isnull(),0,df_a["release_year"])

In [109]:
df_a

song       artist release_year  \
0                                Caught Up in You  .38 Special         1982   
1                                    Fantasy Girl  .38 Special            0   
2                                 Hold On Loosely  .38 Special         1981   
3                          Rockin' Into the Night  .38 Special         1980   
4                               Art For Arts Sake         10cc         1975   
...                                           ...          ...          ...   
2225                      She Loves My Automobile       ZZ Top            0   
2226                            Tube Snake Boogie       ZZ Top         1981   
2227                                         Tush       ZZ Top         1975   
2228                                   TV Dinners       ZZ Top         1983   
2229  WAITIN' FOR THE BUS/JESUS JUST LEFT CHICAGO       ZZ Top         1973   

                                            song_artist  first_flag  \
0                       Caught Up in You by .38 Special           1   
1                           Fantasy Girl by .38 Special           1   
2                        Hold On Loosely by .38 Special           1   
3                 Rockin' Into the Night by .38 Special           1   
4                             Art For Arts Sake by 10cc           1   
...                                                 ...         ...   
2225                  She Loves My Automobile by ZZ Top           1   
2226                        Tube Snake Boogie by ZZ Top           1   
2227                                     Tush by ZZ Top           1   
2228                               TV Dinners by ZZ Top           1   
2229  WAITIN' FOR THE BUS/JESUS JUST LEFT CHICAGO by...           1   

      year_flag  playcount  year_x_playcount  
0             1         82                82  
1             0          3                 0  
2             1         85                85  
3             1         18                18  
4             1          1                 1  
...         ...        ...               ...  
2225          0          1                 0  
2226          1         32                32  
2227          1        109               109  
2228          1          1                 1  
2229          1          2                 2  

[2230 rows x 8 columns]

In [130]:
df_a.loc[df_a["release_year"].isnull(),"release_year"]=0

In [131]:
df_a

song       artist release_year  \
0                                Caught Up in You  .38 Special         1982   
1                                    Fantasy Girl  .38 Special            0   
2                                 Hold On Loosely  .38 Special         1981   
3                          Rockin' Into the Night  .38 Special         1980   
4                               Art For Arts Sake         10cc         1975   
...                                           ...          ...          ...   
2225                      She Loves My Automobile       ZZ Top            0   
2226                            Tube Snake Boogie       ZZ Top         1981   
2227                                         Tush       ZZ Top         1975   
2228                                   TV Dinners       ZZ Top         1983   
2229  WAITIN' FOR THE BUS/JESUS JUST LEFT CHICAGO       ZZ Top         1973   

                                            song_artist  first_flag  \
0                       Caught Up in You by .38 Special           1   
1                           Fantasy Girl by .38 Special           1   
2                        Hold On Loosely by .38 Special           1   
3                 Rockin' Into the Night by .38 Special           1   
4                             Art For Arts Sake by 10cc           1   
...                                                 ...         ...   
2225                  She Loves My Automobile by ZZ Top           1   
2226                        Tube Snake Boogie by ZZ Top           1   
2227                                     Tush by ZZ Top           1   
2228                               TV Dinners by ZZ Top           1   
2229  WAITIN' FOR THE BUS/JESUS JUST LEFT CHICAGO by...           1   

      year_flag  playcount  year_x_playcount  
0             1         82                82  
1             0          3                 0  
2             1         85                85  
3             1         18                18  
4             1          1                 1  
...         ...        ...               ...  
2225          0          1                 0  
2226          1         32                32  
2227          1        109               109  
2228          1          1                 1  
2229          1          2                 2  

[2230 rows x 8 columns]

#### 4.B Verify that `release` contains no null values.

In [ ]:
# your answer here

In [118]:
df_a["release_year"].value_counts()

0                577
1973             104
1975              83
1977              83
1970              81
1971              75
1969              72
1980              70
1978              64
1979              63
1967              61
1981              61
1983              60
1976              56
1982              54
1984              51
1972              50
1974              48
1968              46
1985              39
1987              39
1986              37
1991              34
1989              32
1966              30
1988              29
1965              28
1994              25
1990              22
1993              19
1964              14
1992              14
1999              13
1995              10
1997               9
1996               9
1963               9
1998               6
2002               6
2004               5
2005               5
2012               5
2001               4
2011               3
2007               3
2008               3
2003               3
2000         

### 5. Ensure that the data types of the columns make sense. 

Verifying column data types is a critical part of data munging. If columns have the wrong data type, then there is usually corrupted or incorrect data in some of the observations.

#### 5.A Look at the data types for the columns. Are any incorrect given what the data represents?

In [ ]:
# your answer here

In [120]:
df_a.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   song              2230 non-null   object
 1   artist            2230 non-null   object
 2   release_year      2230 non-null   object
 3   song_artist       2230 non-null   object
 4   first_flag        2230 non-null   int64 
 5   year_flag         2230 non-null   int64 
 6   playcount         2230 non-null   int64 
 7   year_x_playcount  2230 non-null   int64 
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


In [ ]:
# I would expect release_year to be an int, 
# also a scan of the values shows songkick.com

_Only the `release` column appears to be wrong. It is represented as a string but should be an integer for year._

### 6. Investigate and clean up the `release` column.

The `release` column is a string data type when it should be an integer.

#### 6.A Figure out what value(s) are causing the `release` column to be encoded as a string instead of an integer.

In [ ]:
# your answer here

In [138]:
df_a[df_a["release_year"].str.isdigit()==False]

song          artist   release_year  \
1504  Bullfrog Blues  Rory Gallagher  SONGFACTS.COM   

                           song_artist  first_flag  year_flag  playcount  \
1504  Bullfrog Blues by Rory Gallagher           1          1          1   

      year_x_playcount  
1504                 1

In [141]:
df_a[df_a["release_year"].str.isdigit()==False].index

Int64Index([1504], dtype='int64')

#### 6.B Look at the rows in which there is incorrect data in the `release` column.

In [ ]:
# your answer here

In [150]:
df_a.loc[1504]

song                                  Bullfrog Blues
artist                                Rory Gallagher
release_year                           SONGFACTS.COM
song_artist         Bullfrog Blues by Rory Gallagher
first_flag                                         1
year_flag                                          1
playcount                                          1
year_x_playcount                                   1
Name: 1504, dtype: object

'SONGFACTS.COM'

#### 6.C. Clean up the data. Normally we may replace the offending data with null np.nan values, however we previously converted all of the nan values in the release column to zeros so we might as well continue with the same practice. Replacing with 0 (or nan) will allow us to convert the column to numeric.

In [ ]:
# your answer here

In [154]:
df_a.loc[1504,"release_year"]

'SONGFACTS.COM'

In [155]:
df_a.loc[1504,"release_year"] = 0

In [158]:
df_a.loc[:,"release_year"] = np.int64(df_a.loc[:,"release_year"])

In [159]:
df_a.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   song              2230 non-null   object
 1   artist            2230 non-null   object
 2   release_year      2230 non-null   int64 
 3   song_artist       2230 non-null   object
 4   first_flag        2230 non-null   int64 
 5   year_flag         2230 non-null   int64 
 6   playcount         2230 non-null   int64 
 7   year_x_playcount  2230 non-null   int64 
dtypes: int64(5), object(3)
memory usage: 139.5+ KB


In [161]:
df_a["release_year"].value_counts()

0       578
1973    104
1975     83
1977     83
1970     81
1971     75
1969     72
1980     70
1978     64
1979     63
1967     61
1981     61
1983     60
1976     56
1982     54
1984     51
1972     50
1974     48
1968     46
1985     39
1987     39
1986     37
1991     34
1989     32
1966     30
1988     29
1965     28
1994     25
1990     22
1993     19
1992     14
1964     14
1999     13
1995     10
1996      9
1963      9
1997      9
1998      6
2002      6
2004      5
2012      5
2005      5
2001      4
2000      3
2008      3
2011      3
1962      3
2003      3
2007      3
2013      2
2014      2
2006      1
1071      1
1955      1
1961      1
1958      1
Name: release_year, dtype: int64

**Note:** Year can also be considered a descriptive value and therefore it also makes sense for a Year column to be an object.  
However, just like in this situation, using conversion to numerics is a great way of identifying improper values in a year column.

### 7. Get summary statistics for the `release` column using the `.describe()` function.

Now that the `release` column is finally a numeric data type, we can apply the `.describe()` function.  

#### 7.A Print out the summary stats for the `release` column. What is the earliest and latest release date?

In [ ]:
# your answer here

In [162]:
df_a["release_year"].describe()

count    2230.000000
mean     1465.331390
std       867.196161
min         0.000000
25%         0.000000
50%      1973.000000
75%      1981.000000
max      2014.000000
Name: release_year, dtype: float64

In [ ]:
# min 0 - due to replacement
# max 2014

In [166]:
df_a[(df_a["release_year"]>0)].describe()

release_year  first_flag  year_flag    playcount  year_x_playcount
count   1652.000000      1652.0     1652.0  1652.000000       1652.000000
mean    1978.019976         1.0        1.0    20.312954         20.312954
std       24.191247         0.0        0.0    27.502560         27.502560
min     1071.000000         1.0        1.0     0.000000          0.000000
25%     1971.000000         1.0        1.0     2.000000          2.000000
50%     1977.000000         1.0        1.0     7.000000          7.000000
75%     1984.000000         1.0        1.0    28.000000         28.000000
max     2014.000000         1.0        1.0   142.000000        142.000000

#### 7.B Based on the summary statistics, is there anything else wrong with the `release` column? 

In [ ]:
# your answer here

In [169]:
df_a[(df_a["release_year"]<=1900)&(df_a["release_year"]>0)]

song      artist  release_year          song_artist  first_flag  \
547  Levon  Elton John          1071  Levon by Elton John           1   

     year_flag  playcount  year_x_playcount  
547          1          8                 8

_Looking at the DataFrame that contains the year 1071, we can see that the year was probably corrupted and should be replaced with something else if possible._

In [172]:
df_a.loc[547,"release_year"]=1971

In [173]:
df_a.loc[547,]

song                              Levon
artist                       Elton John
release_year                       1971
song_artist         Levon by Elton John
first_flag                            1
year_flag                             1
playcount                             8
year_x_playcount                      8
Name: 547, dtype: object

### 8. Make changes and investigate using custom functions with `.apply()`.

Let's say we want to traverse every single row in our data set and apply a function to that row.

#### 8.A Write a function that will take a row of a DataFrame and print out the song, artist, and whether or not the release date is < 1970.


In [ ]:
# your answer here

In [178]:
def key_info(targ_row):
    if targ_row["release_year"] < 1970:
        release_date_check = "Before 1970"
    else:
        release_date_check = "After 1970"
    print(f"Song: {targ_row['song']} | Artist: {targ_row['song']} | Release year: {release_date_check}.")
    return

In [179]:
key_info(df_a.loc[547,])

Song: Levon | Artist: Levon | Release year: After 1970.


#### 8.B Using the `.apply()` function, apply the function you wrote to the first four rows of the DataFrame.

You will need to tell the `apply` function to operate row by row. Setting the keyword argument as `axis=1` indicates that the function should be applied to each row individually.

In [ ]:
# your answer here

In [181]:
?df_a.apply

In [188]:
df_a.loc[:4,:]

song       artist  release_year  \
0        Caught Up in You  .38 Special          1982   
1            Fantasy Girl  .38 Special             0   
2         Hold On Loosely  .38 Special          1981   
3  Rockin' Into the Night  .38 Special          1980   
4       Art For Arts Sake         10cc          1975   

                             song_artist  first_flag  year_flag  playcount  \
0        Caught Up in You by .38 Special           1          1         82   
1            Fantasy Girl by .38 Special           1          0          3   
2         Hold On Loosely by .38 Special           1          1         85   
3  Rockin' Into the Night by .38 Special           1          1         18   
4              Art For Arts Sake by 10cc           1          1          1   

   year_x_playcount  
0                82  
1                 0  
2                85  
3                18  
4                 1

In [193]:
df_a.loc[:3,:].apply(key_info,axis=1)

Song: Caught Up in You | Artist: Caught Up in You | Release year: After 1970.
Song: Fantasy Girl | Artist: Fantasy Girl | Release year: Before 1970.
Song: Hold On Loosely | Artist: Hold On Loosely | Release year: After 1970.
Song: Rockin' Into the Night | Artist: Rockin' Into the Night | Release year: After 1970.


0    None
1    None
2    None
3    None
dtype: object

You'll notice that there will be a final output Series of `None` values. The `.apply()` function, if a return value is not specified, will return a Series of `None` values (similar to how the default return for Python functions is `None` when a return statement is not specified).